src/lightning_app/core/api.py

```diff
@@ -172,33 +179,71 @@ async def get_state(
     if x_lightning_session_id is None:
         raise Exception("Missing X-Lightning-Session-ID header")
 
+    if not ENABLE_PULLING_STATE_ENDPOINT:
+        response.status_code = status.HTTP_405_METHOD_NOT_ALLOWED
+        return {"status": "failure", "reason": "This endpoint is disabled."}
+
     with lock:
         x_lightning_session_uuid = TEST_SESSION_UUID
         state = global_app_state_store.get_app_state(x_lightning_session_uuid)
         global_app_state_store.set_served_state(x_lightning_session_uuid, state)
         return state
 
 
+def get_component_by_name(component_name: str, state):
+    child = state
+    for child_name in component_name.split(".")[1:]:
+        try:
+            child = child["flows"][child_name]
+        except KeyError:
+            child = child["structures"][child_name]
+
+    if isinstance(child["vars"]["_layout"], list):
+        assert len(child["vars"]["_layout"]) == 1
+        return child["vars"]["_layout"][0]["target"]
+    return child["vars"]["_layout"]["target"]
+
+
+@fastapi_service.get("/api/v1/layout", response_class=JSONResponse)
+async def get_layout() -> Mapping:
+    with lock:
+        x_lightning_session_uuid = TEST_SESSION_UUID
+        state = global_app_state_store.get_app_state(x_lightning_session_uuid)
+        global_app_state_store.set_served_state(x_lightning_session_uuid, state)
+        layout = deepcopy(state["vars"]["_layout"])
+        for la in layout:
+            if la["content"].startswith("root."):
+                la["content"] = get_component_by_name(la["content"], state)
+        return layout
+
+
 @fastapi_service.get("/api/v1/spec", response_class=JSONResponse)
 async def get_spec(
+    response: Response,
     x_lightning_session_uuid: Optional[str] = Header(None),
     x_lightning_session_id: Optional[str] = Header(None),
-) -> List:
+) -> Union[List, Dict]:
     if x_lightning_session_uuid is None:
         raise Exception("Missing X-Lightning-Session-UUID header")
     if x_lightning_session_id is None:
         raise Exception("Missing X-Lightning-Session-ID header")
+
+    if not ENABLE_PULLING_STATE_ENDPOINT:
+        response.status_code = status.HTTP_405_METHOD_NOT_ALLOWED
+        return {"status": "failure", "reason": "This endpoint is disabled."}
+
     global app_spec
     return app_spec or []
 
 
 @fastapi_service.post("/api/v1/delta")
```

Comment:
Should `/api/v1/upload_file/{filename}` route be also disabled ?

Reply:
+1

Reply:
Done ! I didn't do it at first as this endpoint would be removed as soon as upload pre-signed url is available.